Stabilize BLS signatures and implement host functions

s0me0ne-unkn0wn opened the pull request:
This PR stabilizes BLS signatures by removing `bls-experimental` feature gate and implements [RFC-156](https://github.com/polkadot-fellows/RFCs/pull/156).

It uses [RFC-145](https://github.com/polkadot-fellows/RFCs/pull/145) implementation branch as the base branch, as the implementations have to be bundled according to #9876.

s0me0ne-unkn0wn committed:
Stabilize BLS signatures and implement host functions

davxy commented on substrate/primitives/io/src/lib.rs:
```diff
@@ -2712,55 +2711,289 @@ pub trait Crypto {
 		id: PassPointerAndReadCopy<KeyTypeId, 4>,
 		seed: PassFatPointerAndDecode<Option<Vec<u8>>>,
-	) -> AllocateAndReturnPointer<bls381::Public, 144> {
+		out: PassPointerAndWrite<&mut bls381::Public, 144>,
```
Why did you choose to pass the output buffer instead of returning it by value?
It may be functionally equivalent, but this API design differs from the other "generate" methods, which return a value. I would prefer a consistent design.

davxy reviewed (commented):
Technically this looks fine, as you are only removing the `bls-experimental` feature.

My main concern is the target of this PR. I don't think it should be bundled with the PR for [RFC-4](https://github.com/polkadot-fellows/RFCs/pull/4)/[RFC-145](https://github.com/polkadot-fellows/RFCs/pull/145), which covers a completely different topic. If we proceed this way, merging the PR for RFC-4 would also include the BLS stabilization on master branch. I think these changes should be introduced upstream  in a dedicated PR, and only after [RFC-156](https://github.com/polkadot-fellows/RFCs/pull/156) is merged.

s0me0ne-unkn0wn commented on substrate/primitives/io/src/lib.rs:
```diff
@@ -2712,55 +2711,289 @@ pub trait Crypto {
+
+	/// Returns the `ecdsa & bls12-381` public key for the given key type and index in the keystore.
+	/// Panics if the key index is out of bounds.
+	fn ecdsa_bls381_public_key(
```
> So now there is a dedicated hostcall to fetch a single public key and to count the public keys?

Yes, it follows the logic introduced by RFC-145: https://github.com/s0me0ne-unkn0wn/RFCs/blob/rfc4/text/0145-remove-unnecessary-allocator-usage.md#ext_crypto_ed25519sr25519ecdsa_num_public_keys

The old behavior ("give me all the public keys") is obsolete, as the caller doesn't know the size of the buffer that it should provide to store all the keys.

To make the transition smoother, I implemented wrappers. A wrapper is basically a front-end code that only exists on the runtime side. It provides the old familiar interface to the developer, and inside it uses the new host functions as a backend to perform actual host calls.

In this case, we're exposing both new hostcalls (*_num_public_keys and *_public_key) and a wrapper (*_public_keys). Developers are free to use any combination of them.

However, we cannot exclude the possibility that this public key fetching interface may change during the RFC-145 discussion (I'm personally not entirely satisfied with its design).

drskalman commented on substrate/primitives/io/src/lib.rs:
```diff
@@ -2712,55 +2711,289 @@ pub trait Crypto {
+	/// Stores the signature in the provided output buffer.
+	/// Returns 0 on success, -1 on error.
+	#[wrapped]
+	fn ecdsa_bls381_sign_with_keccak256(
```
same here, bls doesn't sign with sha3 it is only ecdsa so we can simulate the action with other functions.

drskalman reviewed (changes requested):
I'm also concern about extra host functions. The RFC at least does not mandates that every crypto type needs a sign function right?

drskalman committed:
Remove owner from pop rename pop keyproofs (#11369)

This PR:
- changes `bls381_generate_proof_of_possession` to actually only
generate proof of possesions (not both proof of possession and
ownership).
- Rename ProofOfPossession to KeyProofs in all traits because that is
more correct naming, because ProofOfPossession is only part of
KeyProofs.

as a result:
- for all crypto types except for bls, key proofs is only ownership
proof.
- for bls it is both owneship proof and proof of possession.

s0me0ne-unkn0wn committed:
Merge remote-tracking branch 'origin/s0me0ne/rfc4' into s0me0ne/rfc156

s0me0ne-unkn0wn committed:
RFC compliance

s0me0ne-unkn0wn committed:
Merge remote-tracking branch 'origin/s0me0ne/rfc4' into s0me0ne/rfc156
